### PR TITLE
pg_repack extension created for staging commcarehq_synclog

### PR DESCRIPTION
**Pull request by prem-fissionhq**
Installed pg_repack 1.4.3 version on staging pg_bouncer server to add an extension on the staging commcarehq_synclog database. Faced some issues in pg_repack tasks/main.yml with the item variable. Fixed it. 
- Updated pg_repack script to connect to the remote database. 
- we need superuser (root) credentials to allow repacking on the databases/tables. 

Please review my changes related pg_repack.py. If looks good will apply the changes in staging env first. thanks.

### CODE REVIEW

**Comment by prem-fissionhq** [comment]
@snopoke please review my changes related to `pg_repack.py` script and let me your suggestions. thanks

File: src/commcare_cloud/ansible/roles/pg_repack/defaults/main.yml [hunk 1]
```diff
@@ -1,3 +1,5 @@
 pg_repack_version: 1.4.5
 pg_repack_source: "http://api.pgxn.org/dist/pg_repack/{{pg_repack_version}}/pg_repack-{{pg_repack_version}}.zip"
 pg_repack_script_path: /usr/local/bin/pg_repack.py
+nadir_hour: 2
+pg_bouncer_port: 6432
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/pg_repack/defaults/main.yml:5]: I Agree. I tried this variable setting from the DB config, it didn't work only one variable was getting considered. I wil cross check and get back to you. 

```
pg_repack:
  pgproxy2:
    - database: commcarehq_synclogs (with and without comma here)
    - port: 6432
```

File: src/commcare_cloud/ansible/roles/pg_repack/defaults/main.yml [hunk 2]
```diff
@@ -1,3 +1,5 @@
 pg_repack_version: 1.4.5
 pg_repack_source: "http://api.pgxn.org/dist/pg_repack/{{pg_repack_version}}/pg_repack-{{pg_repack_version}}.zip"
 pg_repack_script_path: /usr/local/bin/pg_repack.py
+nadir_hour: 2
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/pg_repack/defaults/main.yml:4]: Noted. Sure, will do that.

File: src/commcare_cloud/ansible/roles/pg_repack/defaults/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (snopoke) [reply src/commcare_cloud/ansible/roles/pg_repack/defaults/main.yml:5]: it should look like this:
```yaml
pg_repack:
  pgproxy2:
    - database: commcarehq_synclogs
      port: 6432
```

File: src/commcare_cloud/ansible/roles/pg_repack/tasks/main.yml [hunk 3]
```diff
@@ -46,11 +46,13 @@
...
-      db: "{{ item }}"
-      port: "{{ postgresql_port }}"
+      db: "{{ item.database }}"
+      port: "{{ pg_bouncer_port|default(5432) }}"
+      login_user: "{{ postgres_users.root.username }}"
+      login_password: "{{ postgres_users.root.password }}"
       version: "{{ pg_repack_version }}"
     loop: "{{ pg_repack.get(inventory_hostname, []) }}"
-  when: pg_repack_stat.stat.exists == False or pg_repack_installed_version.stdout.split()[1] != pg_repack_version
+  when: pg_repack_stat.stat.exists == False or pg_repack_installed_version.stdout.split()[1] == pg_repack_version
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/roles/pg_repack/tasks/main.yml:55]: sure.

**Comment by prem-fissionhq** [comment]
@snopoke Updated the files and script as per your suggestions. Could you please review my changes? thank you.

File: src/commcare_cloud/ansible/roles/pg_repack/files/pg_repack.py [hunk 4]
```diff
@@ -146,9 +151,17 @@ def main():
...
 
-    repack_command = [
-        args.pg_repack, '--no-order', '--wait-timeout=30', f'--dbname={args.database}'
-    ] + [f'--table={table}' for table in table_names]
+    repack_command = [args.pg_repack, '--no-order', '--wait-timeout=30', f'--dbname={args.database}' ] + [f'--table={table}' for table in table_names]
+
+    additional_args = {'port':args.port,'username':args.username, 'host':args.host}
+
+    for key, value in additional_args.items():
+        if additional_args[key]:
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/pg_repack/files/pg_repack.py:159]:
```suggestion
        if value:
```

**Review by snopoke — APPROVED** [review]
small comments but looks good otherwise

**Comment by prem-fissionhq** [comment]
> small comments but looks good otherwise

Thanks, Simon. I will make those changes.

**Comment by prem-fissionhq** [comment]
If I defined postgres_users.root.username and postgres_users.root.password variable values in the postgresql.yml to username/password. The user root/commcare is not identified. 
```
pg_repack:
  pgproxy2:
    - database: commcarehq_synclogs
      username: "{{ postgres_users.root.username }}"
      password: "{{ postgres_users.root.password }}"
      port: 6432
      skip_superuser_check: False
```
Error:

ERROR! The field 'hosts' has an invalid value, which includes an undefined variable. The error was: 'unicode object' has no attribute 'root'

The error appears to be in '/Users/premkumar/Downloads/Fission/Dimagi/git/commcare-cloud/src/commcare_cloud/ansible/deploy_postgres.yml': line 85, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: Install and configure pg_repack
  ^ here
✗ Check failed with status code 4

However, when the same strings "{{ postgres_users.root.username }}" and "{{ postgres_users.root.password }}" passed directly into the main.yml of pg_repack it works. Can you please help me here, what i'm missing here. Thanks

**Comment by snopoke** [comment]
@prem-fissionhq you can change it to the following in `deploy_postgres.yml`:
```yaml
- name: Install and configure pg_repack
  hosts: postgresql
  become: true
  roles:
    - role: pg_repack
      when: inventory_hostname in pg_repack
  tags:
    pg_repack
```

**Comment by prem-fissionhq** [comment]
Thanks, Simon. It worked.
All the changes have been made and tested in Staging env, the cron is correctly set too. Please review my changes and let me know if you have any more suggestions. Thanks
cc: @dannyroberts

**Review by snopoke — APPROVED** [review]

**Pull request merged**